tiledb/sm/array_schema/dimension.cc

```diff
@@ -734,24 +750,46 @@ double Dimension::overlap_ratio(const Range& r1, const Range& r2) const {
 template <>
 void Dimension::split_range<char>(
     const Range& r, const ByteVecValue& v, Range* r1, Range* r2) {
-  assert(!r.empty());
   assert(!v.empty());
-  assert(!r.unary());
   assert(r1 != nullptr);
   assert(r2 != nullptr);
 
   // First range
+  auto min_string = std::string("\x0", 1);
+  auto new_r1_start = !r.start_str().empty() ? r.start_str() : min_string;
   auto new_r1_end = std::string((const char*)v.data(), v.size());
-  r1->set_str_range(r.start_str(), new_r1_end);
+  auto new_r1_end_size = (int)new_r1_end.size();
+  int pos;
+  for (pos = 0; pos < new_r1_end_size; ++pos) {
+    if ((int)new_r1_end[pos] < 0) {
+      new_r1_end[pos] = 127;
+      new_r1_end.resize(pos + 1);
+      break;
+    }
+  }
+  assert(pos < new_r1_end_size);  // One negative char must be present
+  r1->set_str_range(new_r1_start, new_r1_end);
 
   // Second range
-  auto start = std::string((const char*)v.data(), v.size());
-  auto end = r.end_str();
-  auto split_pos = utils::parse::common_prefix_size(start, end);
-  auto new_v = (char)(start[split_pos] + 1);
-  auto new_r2_start = start.substr(0, split_pos) + new_v;
+  auto new_r2_start = std::string((const char*)v.data(), v.size());
+  // The following will make "a b -1 -4 0" -> "a c"
+  for (pos = 0; pos < new_r1_end_size; ++pos) {
+    if ((int)new_r2_start[pos] < 0)
```

Comment:
Note that `v` is the splitting value calculated by `Dimension::map_from_uint64`. That value spans all `uint64` values and therefore any individual byte from that value can be in `[0, 255]` or in `[-127, 127]` when cast to `int`. The algorithm finds the first negative byte that corresponds to an ASCII overflow and truncates the string there. Then later it "adds one" to the right of the string.

Reply:
In fact, those `new_{r1,r2}_{start,end}` values are `string` and the underlying elements are (signed) `char`. The cast to `int` preserves the negative sign.

Reply:
You're right -- `char` appears to be signed on our supported machines. However, this is implementation-specific and not in the standard (this can be modified with compiler flags on both gcc and MSVC). It's unlikely to change, but we would be safer here by casting to a `signed char` before casting to an `int`.